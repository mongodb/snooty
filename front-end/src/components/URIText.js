import React from 'react';

const TEMPLATE_TYPE_SELF_MANAGED = 'local MongoDB';
const TEMPLATE_TYPE_REPLICA_SET = 'local MongoDB with replica set';
const TEMPLATE_TYPE_ATLAS_36 = 'Atlas (Cloud) v. 3.6';
const TEMPLATE_TYPE_ATLAS_34 = 'Atlas (Cloud) v. 3.4';
const TEMPLATE_TYPE_ATLAS = 'Atlas (Cloud)';

const URI_PLACEHOLDER = '<URISTRING>';
const USERNAME_PLACEHOLDER = '<USERNAME>';
const URISTRING_SHELL_PLACEHOLDER = '<URISTRING_SHELL>';
const URISTRING_SHELL_NOUSER_PLACEHOLDER = '<URISTRING_SHELL_NOUSER>';

const TEMPLATE_OPTIONS = {
  [TEMPLATE_TYPE_SELF_MANAGED]: [
    {
      name: 'authSource',
      type: 'text'
    }
  ],
  [TEMPLATE_TYPE_REPLICA_SET]: [
    {
      name: 'replicaSet',
      type: 'text'
    },
    {
      name: 'authSource',
      type: 'text'
    },
    {
      name: 'ssl',
      type: 'pass-through',
      value: 'true'
    }
  ],
  [TEMPLATE_TYPE_ATLAS_36]: [],
  [TEMPLATE_TYPE_ATLAS_34]: [
    {
      name: 'replicaSet',
      type: 'text'
    },
    {
      name: 'authSource',
      type: 'text'
    },
    {
      name: 'ssl',
      type: 'pass-through',
      value: 'true'
    }
  ],
}

const URIText = ({ value, templateType, uri }) => {
  return replacePlaceholderWithURI(value, templateType, uri);
}

const replacePlaceholderWithURI = (value, templateType, uri) => {
  return value
    .replace(URI_PLACEHOLDER, generateURI(uri, templateType, 'template'))
    .replace(USERNAME_PLACEHOLDER, uri.username || '$[username]')
    .replace(URISTRING_SHELL_PLACEHOLDER, generateURI(uri, templateType, 'templateShell'))
    .replace(URISTRING_SHELL_NOUSER_PLACEHOLDER, generateURI(uri, templateType, 'templatePasswordRedactedShell'));
}

const optionStringifier = (options, uri) => {
  const paramJoinCharacter = '&';
  let optionParams = [];
  options.forEach(({ name, type, value }) => {
    let paramValue = uri[name];
    if (!paramValue) {
      paramValue = type === 'pass-through' ? value : (uri[name] || `$[${name}]`);
    }
    optionParams.push(`${name}=${paramValue}`)
  });
  return optionParams.join(paramJoinCharacter);
}

const generateURI = (uri, templateName, templateType) => {
  if (uri.env) {
    templateName = uri.env;
  } else if (templateName === TEMPLATE_TYPE_ATLAS) {
    templateName = TEMPLATE_TYPE_ATLAS_34;
  }

  const username = uri.username || '$[username]';
  const authSource = uri.authSource || '$[authSource]';
  const database = uri.database || '$[database]';
  const replicaSet = uri.replicaSet || '$[replicaSet]';
  const options = optionStringifier(TEMPLATE_OPTIONS[templateName], uri);
  let hostlist = '$[hostlist]';
  if (uri.hostlist && Object.values(uri.hostlist).length > 0) {
    hostlist = Object.values(uri.hostlist);
    hostlist.filter(host => host !== '');
  }

  const TEMPLATES = {
    [TEMPLATE_TYPE_SELF_MANAGED]: {
      template: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
      templatePasswordRedactedShell: `mongodb://${hostlist}/${database}?${options} --username ${username}`,
      templateShell: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
    },
    [TEMPLATE_TYPE_REPLICA_SET]: {
      template: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
      templatePasswordRedactedShell: `mongodb://${hostlist}/${database}?${options} --username ${username} --password`,
      templateShell: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
    },
    [TEMPLATE_TYPE_ATLAS_36]: {
      template: `mongodb://${username}:$[password]@${hostlist}/${database}?retryWrites=true`,
      templatePasswordRedactedShell: `mongodb+srv://${hostlist}/${database} --username ${username} --password`,
      templateShell: `mongodb+srv://${username}:$[password]@${hostlist}/${database}`,
    },
    [TEMPLATE_TYPE_ATLAS_34]: {
        template: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
        templatePasswordRedactedShell: `mongodb://${hostlist}/${database}?replicaSet=${replicaSet} --ssl --authenticationDatabase ${authSource} --username ${username} --password`,
        templateShell: `mongodb://${hostlist}/${database}?replicaSet=${replicaSet} --ssl --authenticationDatabase ${authSource} --username ${username} --password $[password]`,
    }
  };

  return TEMPLATES[templateName][templateType];
}

export default URIText;
