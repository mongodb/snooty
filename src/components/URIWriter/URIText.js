import PropTypes from 'prop-types';
import {
  TEMPLATE_TYPE_SELF_MANAGED,
  TEMPLATE_TYPE_REPLICA_SET,
  TEMPLATE_TYPE_ATLAS_36,
  TEMPLATE_TYPE_ATLAS_34,
  TEMPLATE_TYPE_ATLAS,
  URI_PLACEHOLDER,
  USERNAME_PLACEHOLDER,
  URISTRING_SHELL_PLACEHOLDER,
  URISTRING_SHELL_NOUSER_PLACEHOLDER,
} from './constants';

const TEMPLATE_OPTIONS = {
  [TEMPLATE_TYPE_SELF_MANAGED]: [
    {
      name: 'authSource',
      type: 'text',
    },
  ],
  [TEMPLATE_TYPE_REPLICA_SET]: [
    {
      name: 'replicaSet',
      type: 'text',
    },
    {
      name: 'authSource',
      type: 'text',
    },
    {
      name: 'ssl',
      type: 'pass-through',
      value: 'true',
    },
  ],
  [TEMPLATE_TYPE_ATLAS_36]: [],
  [TEMPLATE_TYPE_ATLAS_34]: [
    {
      name: 'replicaSet',
      type: 'text',
    },
    {
      name: 'authSource',
      type: 'text',
    },
    {
      name: 'ssl',
      type: 'pass-through',
      value: 'true',
    },
  ],
};

function optionStringifier(options, uri) {
  const paramJoinCharacter = '&';
  const optionParams = [];
  options.forEach(({ name, type, value }) => {
    let paramValue = uri[name];
    if (!paramValue) {
      paramValue = type === 'pass-through' ? value : uri[name] || `$[${name}]`;
    }
    optionParams.push(`${name}=${paramValue}`);
  });
  return optionParams.join(paramJoinCharacter);
}

function generateURI(uri, activeDeployment, templateType) {
  let template = activeDeployment;
  if (Object.entries(uri).length === 0 && activeDeployment === TEMPLATE_TYPE_ATLAS) {
    template = TEMPLATE_TYPE_ATLAS_36;
  } else if (Object.entries(uri).length === 0 && activeDeployment === TEMPLATE_TYPE_SELF_MANAGED) {
    template = TEMPLATE_TYPE_SELF_MANAGED;
  } else if (uri.atlasVersion) {
    template = uri.atlasVersion;
  } else if (uri.localEnv) {
    template = uri.localEnv;
  }

  const username = uri.username || '$[username]';
  const authSource = uri.authSource || '$[authSource]';
  const database = uri.database || '$[database]';
  const replicaSet = uri.replicaSet || '$[replicaSet]';
  const options = optionStringifier(TEMPLATE_OPTIONS[template], uri);
  const hostlist = uri.hostlist && uri.hostlist.length > 0 ? uri.hostlist : '$[hostlist]';

  const TEMPLATES = {
    [TEMPLATE_TYPE_SELF_MANAGED]: {
      template: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
      templatePasswordRedactedShell: `mongodb://${hostlist}/${database}?${options} --username ${username}`,
      templateShell: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
    },
    [TEMPLATE_TYPE_REPLICA_SET]: {
      template: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
      templatePasswordRedactedShell: `mongodb://${hostlist}/${database}?${options} --username ${username} --password`,
      templateShell: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
    },
    [TEMPLATE_TYPE_ATLAS_36]: {
      template: `mongodb+srv://${username}:$[password]@${hostlist}/${database}?retryWrites=true`,
      templatePasswordRedactedShell: `mongodb+srv://${hostlist}/${database} --username ${username} --password`,
      templateShell: `mongodb+srv://${username}:$[password]@${hostlist}/${database}`,
    },
    [TEMPLATE_TYPE_ATLAS_34]: {
      template: `mongodb://${username}:$[password]@${hostlist}/${database}?${options}`,
      templatePasswordRedactedShell: `mongodb://${hostlist}/${database}?replicaSet=${replicaSet} --ssl --authenticationDatabase ${authSource} --username ${username} --password`,
      templateShell: `mongodb://${hostlist}/${database}?replicaSet=${replicaSet} --ssl --authenticationDatabase ${authSource} --username ${username} --password $[password]`,
    },
  };

  return TEMPLATES[template][templateType];
}

function replacePlaceholderWithURI(value, activeDeployment, uri) {
  return value
    .replace(URI_PLACEHOLDER, generateURI(uri, activeDeployment, 'template'))
    .replace(USERNAME_PLACEHOLDER, uri.username || '$[username]')
    .replace(URISTRING_SHELL_PLACEHOLDER, generateURI(uri, activeDeployment, 'templateShell'))
    .replace(URISTRING_SHELL_NOUSER_PLACEHOLDER, generateURI(uri, activeDeployment, 'templatePasswordRedactedShell'));
}

function URIText({ value, activeDeployment, uriData }) {
  return replacePlaceholderWithURI(value, activeDeployment, uriData);
}

URIText.propTypes = {
  activeDeployment: PropTypes.string,
  uriData: PropTypes.shape({
    atlasVersion: PropTypes.string,
    authSource: PropTypes.string,
    database: PropTypes.string,
    localEnv: PropTypes.string,
    hostlist: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]),
    replicaSet: PropTypes.string,
    username: PropTypes.string,
  }),
  value: PropTypes.string.isRequired,
};

URIText.defaultProps = {
  activeDeployment: TEMPLATE_TYPE_ATLAS,
  uriData: {},
};

export default URIText;
