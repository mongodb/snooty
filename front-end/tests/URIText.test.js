import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import URIText from '../src/components/URIText';

describe('local MongoDB', () => {
  const templateType = 'local MongoDB'
  const uri = {
    authSource: 'admin',
    database: 'myDatabase',
    env: '',
    hostlist: {
      host0: 'cluster0-shard-00-00-igkvv.mongodb.net:27017',
      host1: 'cluster0-shard-00-01-igkvv.mongodb.net:27017',
      host2: 'cluster0-shard-00-02-igkvv.mongodb.net:27017',
    },
    replicaSet: '',
    ssl: '',
    username: 'myUsername',
  };

  describe('when called without a placeholder', () => {
    it('returns the original string', () => {
      const value = `This is the body text`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when there is a <URISTRNG> placeholder', () => {
    it('replaces the placeholder', () => {
      const value = `return MongoClient("<URISTRING>")`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('when there are multiple placeholders', () => {
    it('replaces all placeholders', () => {
      const value =
        `return MongoClient("<URISTRING>")
        username is <USERNAME>
        shell string (<URISTRING_SHELL>)
        without password (<URISTRING_SHELL_NOUSER>)`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});

describe('local MongoDB with replica set', () => {
  const templateType = 'local MongoDB'
  const uri = {
    authSource: '',
    database: 'myDatabase',
    env: 'local MongoDB with replica set',
    hostlist: {
      host0: 'cluster0-shard-00-00-igkvv.mongodb.net:27017',
    },
    replicaSet: 'Cluster0-shard-8',
    ssl: '',
    username: 'myUsername',
  };

  describe('when called without a placeholder', () => {
    it('returns the original string', () => {
      const value = `This is the body text`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when there is a <URISTRNG> placeholder', () => {
    it('replaces the placeholder', () => {
      const value = `return MongoClient("<URISTRING>")`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('when there are multiple placeholders', () => {
    it('replaces all placeholders', () => {
      const value =
        `return MongoClient("<URISTRING>")
        username is <USERNAME>
        shell string (<URISTRING_SHELL>)
        without password (<URISTRING_SHELL_NOUSER>)`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});

describe('Cloud (unspecified version)', () => {
  const templateType = 'Atlas (Cloud)'
  const uri = {
    authSource: 'admin',
    database: 'myDatabase',
    env: '',
    hostlist: {
      host0: 'cluster0-shard-00-00-igkvv.mongodb.net:27017',
    },
    replicaSet: 'Cluster0-shard-8',
    ssl: '',
    username: 'myUsername',
  };

  describe('when called without a placeholder', () => {
    it('returns the original string', () => {
      const value = `This is the body text`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when there is a <URISTRNG> placeholder', () => {
    it('replaces the placeholder', () => {
      const value = `return MongoClient("<URISTRING>")`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('when there are multiple placeholders', () => {
    it('replaces all placeholders', () => {
      const value =
        `return MongoClient("<URISTRING>")
        username is <USERNAME>
        shell string (<URISTRING_SHELL>)
        without password (<URISTRING_SHELL_NOUSER>)`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});

describe('Cloud (MongoDB version 3.4)', () => {
  const templateType = 'Atlas (Cloud)'
  const uri = {
    authSource: 'admin',
    database: 'myDatabase',
    env: 'Atlas (Cloud) v. 3.4',
    hostlist: {
      host0: 'cluster0-shard-00-00-igkvv.mongodb.net:27017',
      host1: 'cluster0-shard-00-01-igkvv.mongodb.net:27017',
      host2: 'cluster0-shard-00-02-igkvv.mongodb.net:27017',
    },
    replicaSet: '',
    ssl: '',
    username: 'myUsername',
  };

  describe('when called without a placeholder', () => {
    it('returns the original string', () => {
      const value = `This is the body text`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when there is a <URISTRNG> placeholder', () => {
    it('replaces the placeholder', () => {
      const value = `return MongoClient("<URISTRING>")`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('when there are multiple placeholders', () => {
    it('replaces all placeholders', () => {
      const value =
        `return MongoClient("<URISTRING>")
        username is <USERNAME>
        shell string (<URISTRING_SHELL>)
        without password (<URISTRING_SHELL_NOUSER>)`;
      const tree = renderer
        .create(<URIText value={value} templateType={templateType} uri={uri} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
