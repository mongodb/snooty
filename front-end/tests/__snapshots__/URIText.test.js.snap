// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud (MongoDB version 3.4) when called without a placeholder returns the original string 1`] = `"This is the body text"`;

exports[`Cloud (MongoDB version 3.4) when there are multiple placeholders replaces all placeholders 1`] = `
"return MongoClient(\\"mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017,cluster0-shard-00-01-igkvv.mongodb.net:27017,cluster0-shard-00-02-igkvv.mongodb.net:27017/myDatabase?replicaSet=$[replicaSet]&authSource=admin&ssl=true\\")
        username is myUsername
        shell string (mongodb://cluster0-shard-00-00-igkvv.mongodb.net:27017,cluster0-shard-00-01-igkvv.mongodb.net:27017,cluster0-shard-00-02-igkvv.mongodb.net:27017/myDatabase?replicaSet=$[replicaSet] --ssl --authenticationDatabase admin --username myUsername --password $[password])
        without password (mongodb://cluster0-shard-00-00-igkvv.mongodb.net:27017,cluster0-shard-00-01-igkvv.mongodb.net:27017,cluster0-shard-00-02-igkvv.mongodb.net:27017/myDatabase?replicaSet=$[replicaSet] --ssl --authenticationDatabase admin --username myUsername --password)"
`;

exports[`Cloud (MongoDB version 3.4) when there is a <URISTRNG> placeholder replaces the placeholder 1`] = `"return MongoClient(\\"mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017,cluster0-shard-00-01-igkvv.mongodb.net:27017,cluster0-shard-00-02-igkvv.mongodb.net:27017/myDatabase?replicaSet=$[replicaSet]&authSource=admin&ssl=true\\")"`;

exports[`Cloud (unspecified version) when called without a placeholder returns the original string 1`] = `"This is the body text"`;

exports[`Cloud (unspecified version) when there are multiple placeholders replaces all placeholders 1`] = `
"return MongoClient(\\"mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017/myDatabase?replicaSet=Cluster0-shard-8&authSource=admin&ssl=true\\")
        username is myUsername
        shell string (mongodb://cluster0-shard-00-00-igkvv.mongodb.net:27017/myDatabase?replicaSet=Cluster0-shard-8 --ssl --authenticationDatabase admin --username myUsername --password $[password])
        without password (mongodb://cluster0-shard-00-00-igkvv.mongodb.net:27017/myDatabase?replicaSet=Cluster0-shard-8 --ssl --authenticationDatabase admin --username myUsername --password)"
`;

exports[`Cloud (unspecified version) when there is a <URISTRNG> placeholder replaces the placeholder 1`] = `"return MongoClient(\\"mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017/myDatabase?replicaSet=Cluster0-shard-8&authSource=admin&ssl=true\\")"`;

exports[`local MongoDB when called without a placeholder returns the original string 1`] = `"This is the body text"`;

exports[`local MongoDB when there are multiple placeholders replaces all placeholders 1`] = `
"return MongoClient(\\"mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017,cluster0-shard-00-01-igkvv.mongodb.net:27017,cluster0-shard-00-02-igkvv.mongodb.net:27017/myDatabase?authSource=admin\\")
        username is myUsername
        shell string (mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017,cluster0-shard-00-01-igkvv.mongodb.net:27017,cluster0-shard-00-02-igkvv.mongodb.net:27017/myDatabase?authSource=admin)
        without password (mongodb://cluster0-shard-00-00-igkvv.mongodb.net:27017,cluster0-shard-00-01-igkvv.mongodb.net:27017,cluster0-shard-00-02-igkvv.mongodb.net:27017/myDatabase?authSource=admin --username myUsername)"
`;

exports[`local MongoDB when there is a <URISTRNG> placeholder replaces the placeholder 1`] = `"return MongoClient(\\"mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017,cluster0-shard-00-01-igkvv.mongodb.net:27017,cluster0-shard-00-02-igkvv.mongodb.net:27017/myDatabase?authSource=admin\\")"`;

exports[`local MongoDB with replica set when called without a placeholder returns the original string 1`] = `"This is the body text"`;

exports[`local MongoDB with replica set when there are multiple placeholders replaces all placeholders 1`] = `
"return MongoClient(\\"mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017/myDatabase?replicaSet=Cluster0-shard-8&authSource=$[authSource]&ssl=true\\")
        username is myUsername
        shell string (mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017/myDatabase?replicaSet=Cluster0-shard-8&authSource=$[authSource]&ssl=true)
        without password (mongodb://cluster0-shard-00-00-igkvv.mongodb.net:27017/myDatabase?replicaSet=Cluster0-shard-8&authSource=$[authSource]&ssl=true --username myUsername --password)"
`;

exports[`local MongoDB with replica set when there is a <URISTRNG> placeholder replaces the placeholder 1`] = `"return MongoClient(\\"mongodb://myUsername:$[password]@cluster0-shard-00-00-igkvv.mongodb.net:27017/myDatabase?replicaSet=Cluster0-shard-8&authSource=$[authSource]&ssl=true\\")"`;
